EX1/HW5.py

```diff
@@ -0,0 +1,118 @@
+import os
+import datetime
+import random
+import re
+
+
+
+def get_files_in_directory(path):
+    file_names = []
+    with os.scandir(path) as directory:
+        for entry in directory:
+            if entry.is_file() and entry.name.startswith("deep"):
+                file_names.append(os.path.basename(entry.path))
+    return file_names
+
+
+def random_date_between_dates(start_date, end_date):
+    start = datetime.datetime.strptime(start_date, '%Y-%m-%d')
+    end = datetime.datetime.strptime(end_date, '%Y-%m-%d')
+    delta = end - start
+    return (start + datetime.timedelta(days=random.randint(0, delta.days))).strftime("%A, %B %d, %Y")
+
+
+def join(*args, sep='-'):
+    result = []
+    for i, lst in enumerate(args):
+        result.extend(lst)
+        if i < len(args) - 1 and sep:
+            result.append(sep)
+    return result
+
+
+def get_recipe_price(prices, optionals=None, **ingredients):
+    total_price = 0
+    for ingredient, amount in ingredients.items():
+        if ingredient not in prices:
+            raise ValueError(f"{ingredient} is not a valid ingredient.")
+        if optionals and ingredient in optionals:
+            continue
+        price_per_100g = prices[ingredient]
+        ingredient_price = (price_per_100g / 100) * amount
+        total_price += ingredient_price
+    return round(total_price, 2)
+
+
+def perfect_numbers():
+    """Generate all perfect numbers."""
+    n = 1
+    while True:
+        divisors = [i for i in range(1, n) if n % i == 0]
+        if sum(divisors) == n:
+            yield n
+        n += 1
+
+
+def extract_messages(filename):  # TODO  fix it.
+    """Extract hidden messages from a binary file."""
+    msg_buffer = ''
+    with open('resources/logo.jpg', 'rb') as f:
+        while True:
+            chunk = f.read(1024)  # Read 1 KB at a time
+            if not chunk:  # End of file
+                break
+            for char in chunk:
+                if char in range(97, 123):  # Check if char is lowercase
+                    msg_buffer += chr(char)
+                    if msg_buffer[-1] == '!':  # Found a complete message
+                        yield msg_buffer[:-1]  # Yield the message without the '!'
+                        msg_buffer = ''
+                else:
+                    msg_buffer = ''  # Reset buffer if non-message char is found
+
+
+def combine_iterables(*iterables):
+    """Combine two or more iterables into tuples."""
+    iters = [iter(it) for it in iterables]
+    while True:
+        try:
+            items = tuple(next(it) for it in iters)
+            yield items
+        except StopIteration:
+            return
+
+
+def modify_chapter_name_and_copy_files(folder_path, new_folder_path):
+    # Create the new directory if it doesn't exist
```

Comment:
Separate code to small functions- 
https://softwareengineering.stackexchange.com/questions/308108/when-is-it-appropriate-to-make-a-separate-function-when-there-will-only-ever-be